metrics/prometheus-metrics-endpoints-and-auto-discovery.md

```diff
@@ -0,0 +1,136 @@
+
+### Non Goals
+* Provide an Analytics service
+* Provide a Mobile SDK for Mobile Developers to store arbitrary metrics
+
+## Terms
+
+- Metrics - Arbitrary measurements that represents current or historical data from a Service, App or User. Some examples:
+  - the current Memory usage
+  - the average number of requests per second
+  - the number of Android users
+  - the total number of in-app purchases for a specific user
+  - Number of failed logins
+- Analytics - The process of applying statistics and maths to data or metrics to answer business-related questions. Some examples:
+  - Are there enough Android users to continue maintaining the Android App?
+  - What type of App users are likely to use in-app purchases?
+
+## Problem Description
+
+During the Mobile Control Panel (MCP) proof of concept (Aug-Nov 2017), the `mcp-standalone` server had functionality added to gather metrics from Mobile Services. This functionality gathered metrics from 2 components:
+
+* Keycloak - by polling the Events API
+* Data Sync - by polling the `/stats` endpoint
+
+The resulting metrics were stored in-memory in the `mcp-standalone` server. Endpoints were exposed to allow the MCP UI to request metrics and render them using Patternfly Widgets e.g. line graphs & counters. In order for the visualisations to be more useful and visually appealing, custom UI code was written for each service, so as to arrange the visualisations in a particular way. For example, queues and processing times were grouped for the Data Sync UI.
+
+This approach surfaced a number of problems:
+
+* Service specific logic required to be written and maintained in the 'mcp-standalone' server
+* Service specific logic required to be written and maintained in the MCP UI
+* Memory footprint of the `mcp-standalone` server was indeterminate
+
+As the number of Mobile Services and supported other Services increases, the size and maintenance overhead of the server & UI would be very large.
+
+Also, post proof of concept, it was decided the `mcp-standalone` server would not be continued as a concept, or any other 'required' server component specific to Mobile.
+
+## Proposed Solution
+
+### Metrics Service
+
+No central metrics gathering or visualisations will be available for Mobile Services 'out of the box'. Instead, if a developer wants metrics, they can provision a Metrics Service from the Service Catalog.
+
+The Metrics Service will consist of:
+
+* Prometheus - for metrics gathering
+* Grafana - for metrics visualisation
+
+Promtehues & Grafana will each have a Persistent Volume for storing gathered metrics and visualisation settings respectively.
+
+Grafana will be configured to have Proemtheus available as a 'Data Source' by default. Data Sources will be stored in a ConfigMap that gets mounted into the Grafana container. e.g.
+
+https://github.com/aerogearcatalog/prometheus-apb/blob/ee86cfc67caaeaabcf7e7bc214c3a6dc6940c333/roles/provision-prometheus-apb/tasks/main.yml#L25-L26
```

Comment:
should we name it different than `prometheus-apb` ?

Reply:
`metrics-apb`?

Reply:
`metrics-apb` sounds good to me.

Reply:
I'll make a note in the proposal to use that name

Reply:
https://issues.jboss.org/browse/AEROGEAR-1819